experimental/include/experimental/Support/BooleanLogic/Parser.h

```diff
@@ -0,0 +1,94 @@
+//===- Parser.h - Parser for boolean expressions ----------------*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the data structure and methods for parsing boolean logic
+// expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef EXPERIMENTAL_SUPPORT_BOOLEANLOGIC_PARSER_H
+#define EXPERIMENTAL_SUPPORT_BOOLEANLOGIC_PARSER_H
+
+#include "experimental/Support/BooleanLogic/BoolExpression.h"
+#include "experimental/Support/BooleanLogic/Lexer.h"
+#include <stack>
+#include <utility>
+
+namespace dynamatic {
+namespace experimental {
+namespace boolean {
+
+// The StackNode struct represents elements in the parsing stack,
+// which can be either expressions or terms (tokens).
+struct StackNode {
+  BoolExpression *expr;
+  Token term;

```

Comment:
With your second constructor this is left (silently) uninitialized. Why not use a `std::optional<Token>` like I suggested before? If somebody ever incorrectly assumes `term` has a value when it's never been initialized explicitly it will crash on the spot with the `std::optional`. With the current implementation it would either crash later on for some hard to track reason, or, even worse, appear to work.